browser/browsing_data/brave_browsing_data_remover_delegate.cc

```diff
@@ -91,3 +110,68 @@ void BraveBrowsingDataRemoverDelegate::ClearShieldsSettings(
     }
   }
 }
+
+#if BUILDFLAG(IPFS_ENABLED)
+void BraveBrowsingDataRemoverDelegate::WaitForIPFSRepoGC(
+    base::Process process) {
+  bool exited = false;
+
+  {
+    base::ScopedAllowBaseSyncPrimitives scoped_allow_base_sync_primitives;
+
+    // Because we set maximum IPFS storage size as 1GB in Brave, ipfs repo gc
+    // command should be finished in just a few seconds and we do not expect
+    // this child process would hang forever. To be safe, we will wait for 30
+    // seconds max here.
+    exited = process.WaitForExitWithTimeout(base::TimeDelta::FromSeconds(30),
+                                            nullptr);
+  }
+
+  if (!exited)
+    process.Terminate(0, false /* wait */);
+}
+
+// Run ipfs repo gc command to clear IPFS cache when IPFS executable path is
+// available. Because the command does not support time ranged cleanup, we will
+// always clear the whole cache expect for pinned files when clearing browsing
+// data.
+void BraveBrowsingDataRemoverDelegate::ClearIPFSCache() {
+  auto* service =
+      ipfs::IpfsServiceFactory::GetInstance()->GetForContext(profile_);
+  if (!service)
+    return;
+
+  base::FilePath path = service->GetIpfsExecutablePath();
+  if (path.empty())
+    return;
+
+  base::CommandLine cmdline(path);
+  cmdline.AppendArg("repo");
+  cmdline.AppendArg("gc");
+
+  base::FilePath data_path = service->GetDataPath();
+  base::LaunchOptions options;
+#if defined(OS_WIN)
+  options.environment[L"IPFS_PATH"] = data_path.value();
+#else
+  options.environment["IPFS_PATH"] = data_path.value();
+#endif
+#if defined(OS_LINUX)
+  options.kill_on_parent_death = true;
+#endif
+#if defined(OS_WIN)
+  options.start_hidden = true;
+#endif
+
+  base::Process process = base::LaunchProcess(cmdline, options);
+  if (!process.IsValid()) {
+    return;
+  }
+
+  base::ThreadPool::PostTaskAndReply(
+      FROM_HERE, {base::TaskPriority::USER_VISIBLE, base::MayBlock()},
+      base::BindOnce(&BraveBrowsingDataRemoverDelegate::WaitForIPFSRepoGC,
+                     base::Unretained(this), base::Passed(&process)),
```

Comment:
(sorry, just randomly came across this)
we really shouldn't use `base::Unretained` like this, because `this` can easily die before the task is started (especially given that most tasks posted around can be time consuming). Note, that `CreateTaskCompletionClosure` binds a weak pointer.

Fortunately, in our case `WaitForIPFSRepoGC` doesn't use that pointer at all, so we shouldn't crash. The most easy fix would be to just post a free function (i.e. not a member of `BraveBrowsingDataRemoverDelegate`) - there are examples of `WaitForExitWithTimeout` across the codebase

Reply:
also, probably we'd better explicitly mention `TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN`

Reply:
Thank you @iefremov, I'll open a follow-up PR for this.